.gitlab/validate-agent-build/validate_agent_build.py

```diff
@@ -0,0 +1,92 @@
+import argparse
+import sys
+import time
+
+import requests
+import os
+
+DATADOG_AGENT_PIPELINE_ID = 22
+BASE_URL = os.environ['CI_API_V4_URL']
+GITLAB_TOKEN = os.environ['GITLAB_TOKEN']
+STAGES_TO_CHECK = ['deps_fetch', 'source_test', 'binary_build', 'package_build']
+
+
+def _get_jobs(pipeline_id, scope=None):
+    all_jobs = []
+    url = f"{BASE_URL}/projects/22/pipelines/{pipeline_id}/jobs?per_page=20"
+    if scope is not None:
+        url += f"&scope={scope}"
+    while True:
+        resp = requests.get(url, headers={'PRIVATE-TOKEN': GITLAB_TOKEN})
+        resp.raise_for_status()
+        all_jobs.extend(resp.json())
+        if 'next' not in resp.links:
+            break
+
+        url = resp.links['next']['url']
+
+    return all_jobs
+
+
+def get_pending_jobs(pipeline_id):
+    scopes = ['created', 'pending', 'running']
+    all_jobs = []
+    for scope in scopes:
+        all_jobs.extend(_get_jobs(pipeline_id, scope=scope))
+
+    jobs = [j for j in all_jobs if j['stage'] in STAGES_TO_CHECK]
+    return jobs
+
+
+def get_failed_jobs(pipeline_id):
+    scopes = ['failed', 'canceled']
+    jobs = []
+    for scope in scopes:
+        jobs.extend(_get_jobs(pipeline_id, scope=scope))
+
+    jobs = [j for j in jobs if not j['allow_failure'] and j['stage'] in STAGES_TO_CHECK]
+    return jobs
+
+
+def retry_failed_jobs(pipeline_id):
+    failed_jobs = get_failed_jobs(pipeline_id)
```

Comment:
I believe failed jobs are always described in the list returned by the jobs API, even if they are retried. Therefore I believe this function will also retry jobs that have failed then passed, which we don't need.

We should filter out jobs that failed at some point, but then succeeded (or are still running because of a retry). To do that, you can get the list of jobs that have the same name as the failed job, and check if at least one instance is in a started / pending / running / success state.

Reply:
Just tested, the api only returns the most recent run of a job unless you specify `include_retried=true` in the URL args.